✨ Added IndexNow integration for faster search engine indexing

betschki opened the pull request:
[IndexNow](https://www.indexnow.org) is a protocol that allows websites to notify search engines (Bing, Yandex, Seznam.cz, Naver) about content changes for faster indexing. This integration automatically pings IndexNow when posts are published or updated, reducing the time for content to appear in search results from days to minutes.

Ghost already integrates Ping-O-Matic, though the actual benefit of that service is rather limited. While IndexNow is not supported by Google, it supports a wider user base than Ping-O-Matic

The feature is behind the 'indexnow' labs flag in developer experiments. API keys are auto-generated on first use and the verification file is served dynamically at /{key}.txt.

---

Got some code for us? Awesome 🎊!

Please take a minute to explain the change you're making:
- Why are you making it?
- What does it do?
- Why is this something Ghost users or developers need?

Please check your PR against these items:

- [x] I've read and followed the [Contributor Guide](https://github.com/TryGhost/Ghost/blob/main/.github/CONTRIBUTING.md)
- [x] I've explained my change
- [x] I've written an automated test to prove my change works

We appreciate your contribution! 🙏

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces IndexNow for faster indexing, gated by `labs.indexnow`.
> 
> - New `core/server/services/indexnow` listens to `post.published` and `post.published.edited`, pings `https://api.indexnow.org/indexnow` when SEO-relevant fields change; initialized in `boot.js`
> - Frontend middleware `serve-indexnow-key` serves the verification file at `/{key}.txt`; mounted in `site.js`
> - New setting `indexnow_api_key` (in default settings), allowed via settings input serializer and excluded from exports
> - Adds `indexnow` to labs flags and Admin Labs UI (with keyed list items)
> - Updates tests: unit tests for service behavior and middleware exposure; snapshot/config hash updated
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8650feb0e7b289b1564a5c9ee32706043f304861. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

betschki commented on ghost/core/core/frontend/web/middleware/serve-indexnow-key.js:
```diff
@@ -0,0 +1,64 @@
+ * theme static assets, so legitimate .txt files in themes take precedence.
+ *
+ * @see ghost/core/core/server/services/indexnow.js - Main IndexNow service
+ */
```
Updated comment to say middleware runs BEFORE theme static assets

betschki commented on ghost/core/core/server/services/indexnow.js:
```diff
@@ -0,0 +1,234 @@
+    const indexNowUrl = new URL(INDEXNOW_ENDPOINT);
+    indexNowUrl.searchParams.set('url', url);
+    indexNowUrl.searchParams.set('key', key);
+    indexNowUrl.searchParams.set('keyLocation', `${siteUrl}${key}.txt`);
```
Fixed. Now using urlUtils.urlJoin() instead of string concatenation

betschki committed:
✨ Added IndexNow integration for faster search engine indexing

IndexNow is a protocol that allows websites to notify search engines
(Bing, Yandex, Seznam.cz, Naver) about content changes for faster
indexing. This integration automatically pings IndexNow when posts are
published or updated, reducing the time for content to appear in search
results from days to minutes.

The feature is behind the 'indexnow' labs flag in developer experiments.
API keys are auto-generated on first use and the verification file is
served dynamically at /{key}.txt.

sagzy commented on ghost/core/core/server/data/migrations/versions/6.10/2025-12-27-11-17-40-add-indexnow-api-key-setting.js:
```diff
@@ -0,0 +1,8 @@
+    value: null,
+    type: 'string',
+    group: 'indexnow'
+});
```
As database migrations have caused pain in the past, we generally prefer to create a separate PR/commit for database migrations, that contains minimal changes (example: https://github.com/TryGhost/Ghost/pull/15705/changes)

Would you mind moving the database changes to a separate PR, so that we can merge the database changes separately from the code changes?

betschki replied:
Makes perfect sense! Created #25808 for that

betschki committed:
Moved API key generation to settings service boot

- Added indexnow_api_key to dynamicDefault in settings model
- Removed getOrCreateApiKey logic from indexnow service
- Added 'coming-soon' to default post slugs to skip
- Separated database migration to PR #25804

betschki committed:
Merge branch 'main' into feature/indexnow-integration

vikaspotluri123 commented on ghost/core/core/server/services/indexnow.js:
```diff
@@ -0,0 +1,212 @@
+        const indexNowUrl = new URL(INDEXNOW_ENDPOINT);
+        indexNowUrl.searchParams.set('url', url);
+        indexNowUrl.searchParams.set('key', key);
+        indexNowUrl.searchParams.set('keyLocation', urlUtils.urlJoin(siteUrl, `${key}.txt`));
```
Yes, but we can use a static route rather than the key (e.g indexnow.txt)
